pulumi convert generates incorrect go code for Terraform's file("") function

### What happened?

Running:
`pulumi convert --from terraform --language go --strict --generate-only `


The `main.tf` has a reference to a file. Here's a code snippet:

```terraform
resource "aws_instance" "example" {
  ami           = "ami-0f3769c8d8429942f"
  instance_type = "t2.micro"             
  tags = {
    Name = "example-instance"
  }
  user_data                   = file("user_data.sh") # <------ BUG OCCURS HERE
  user_data_replace_on_change = true
}
```

### Example

Here's a snippet of the generated `go` code


```go
	example, err := ec2.NewInstance(ctx, "example", &ec2.InstanceArgs{
			Ami:          pulumi.String("ami-0f3769c8d8429942f"),
			InstanceType: pulumi.String("t2.micro"),
			Tags: pulumi.StringMap{
				"name": pulumi.String("example-instance"),
			},
			UserData: std.File(ctx, &std.FileArgs{ // <------ BUG OCCURS HERE
				Input: "user_data.sh",
			}, nil).Result,
			UserDataReplaceOnChange: pulumi.Bool(true),
		})
		if err != nil {
			return err
		}
```

Building the above results in the following error message:

```
go build                                             
# migrate-terraform-pulumi
./main.go:18:14: multiple-value std.File(ctx, &std.FileArgs{…}, nil) (value of type (*std.FileResult, error)) in single-value context
```


The generated go user_data code has an error as std.File returns 2 values but we’re trying to assign a single result to UserData

### Output of `pulumi about`

```pulumi about                                                           
CLI          
Version      3.86.0
Go Version   go1.21.1
Go Compiler  gc

Plugins
NAME  VERSION
aws   6.0.4
go    unknown
std   1.4.0

Host     
OS       darwin
Version  13.5.1
Arch     arm64

This project is written in go: executable='/usr/local/go/bin/go' version='go version go1.21.0 darwin/arm64'
```

### Additional context

Current workaround is to edit the `main.go` code. Following the previous example, we can update the snippet to read:


```go
		userData, err := std.File(ctx, &std.FileArgs{
			Input: "user_data.sh",
		}, nil)
		if err != nil {
			return err
		}
		
		example, err := ec2.NewInstance(ctx, "example", &ec2.InstanceArgs{
			Ami:          pulumi.String("ami-0f3769c8d8429942f"),
			InstanceType: pulumi.String("t2.micro"),
			Tags: pulumi.StringMap{
				"name": pulumi.String("example-instance"),
			},
			UserData:                pulumi.String(userData.Result),
			UserDataReplaceOnChange: pulumi.Bool(true),
		})
		if err != nil {
			return err
		}

```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

--- a/changelog/pending/20231102--programgen-go--fix-using-inline-invoke-expressions-inside-resources-objects-and-arrays.yaml
+++ b/changelog/pending/20231102--programgen-go--fix-using-inline-invoke-expressions-inside-resources-objects-and-arrays.yaml
@@ -1,0 +1,4 @@
+changes:
+- type: fix
+  scope: programgen/go
+  description: Fix using inline invoke expressions inside resources, objects and arrays

--- a/pkg/codegen/go/gen_program.go
+++ b/pkg/codegen/go/gen_program.go
@@ -47,6 +47,7 @@ type generator struct {
 	readDirTempSpiller  *readDirSpiller
 	splatSpiller        *splatSpiller
 	optionalSpiller     *optionalSpiller
+	inlineInvokeSpiller *inlineInvokeSpiller
 	scopeTraversalRoots codegen.StringSet
 	arrayHelpers        map[string]*promptToInputArrayHelper
 	isErrAssigned       bool
@@ -102,6 +103,7 @@ func newGenerator(program *pcl.Program, opts GenerateProgramOptions) (*generator
 		readDirTempSpiller:  &readDirSpiller{},
 		splatSpiller:        &splatSpiller{},
 		optionalSpiller:     &optionalSpiller{},
+		inlineInvokeSpiller: &inlineInvokeSpiller{},
 		scopeTraversalRoots: codegen.NewStringSet(),
 		arrayHelpers:        make(map[string]*promptToInputArrayHelper),
 		externalCache:       opts.ExternalCache,
@@ -1003,6 +1005,10 @@ func (g *generator) genResource(w io.Writer, r *pcl.Resource) {
 			destType, diagnostics := r.InputType.Traverse(hcl.TraverseAttr{Name: input.Name})
 			g.diagnostics = append(g.diagnostics, diagnostics...)
 			expr, temps := g.lowerExpression(input.Value, destType.(model.Type))
+			expr, invokeTemps := g.rewriteInlineInvokes(expr)
+			for _, t := range invokeTemps {
+				temps = append(temps, t)
+			}
 			input.Value = expr
 			g.genTemps(w, temps)
 		}
@@ -1352,6 +1358,12 @@ func (g *generator) genTempsMultiReturn(w io.Writer, temps []interface{}, zeroVa
 			g.Fgenf(w, "}\n")
 		case *optionalTemp:
 			g.Fgenf(w, "%s := %.v\n", t.Name, t.Value)
+		case *inlineInvokeTemp:
+			g.Fgenf(w, "%s, err := %.v\n", t.Name, t.Value)
+			g.Fgenf(w, "if err != nil {\n")
+			g.Fgenf(w, "return err\n")
+			g.Fgenf(w, "}\n")
+			g.isErrAssigned = true
 		default:
 			contract.Failf("unexpected temp type: %v", t)
 		}

--- a/pkg/codegen/go/gen_program_expressions.go
+++ b/pkg/codegen/go/gen_program_expressions.go
@@ -527,7 +527,11 @@ func (g *generator) genObjectConsExpressionWithTypeName(
 		temps = append(temps, kTemps...)
 		item.Key = k
 		x, xTemps := g.lowerExpression(item.Value, item.Value.Type())
+		x, invokeTemps := g.rewriteInlineInvokes(x)
 		temps = append(temps, xTemps...)
+		for _, t := range invokeTemps {
+			temps = append(temps, t)
+		}
 		item.Value = x
 		expr.Items[i] = item
 	}
@@ -748,7 +752,11 @@ func (g *generator) genTupleConsExpression(w io.Writer, expr *model.TupleConsExp
 	var temps []interface{}
 	for i, item := range expr.Expressions {
 		item, itemTemps := g.lowerExpression(item, item.Type())
+		item, invokeTemps := g.rewriteInlineInvokes(item)
 		temps = append(temps, itemTemps...)
+		for _, t := range invokeTemps {
+			temps = append(temps, t)
+		}
 		expr.Expressions[i] = item
 	}
 	g.genTemps(w, temps)
@@ -999,6 +1007,7 @@ func (g *generator) lowerExpression(expr model.Expression, typ model.Type) (
 	expr = pcl.RewritePropertyReferences(expr)
 	expr, diags := pcl.RewriteApplies(expr, nameInfo(0), false /*TODO*/)
 	expr, sTemps, splatDiags := g.rewriteSplat(expr, g.splatSpiller)
+
 	expr, convertDiags := pcl.RewriteConversions(expr, typ)
 	expr, tTemps, ternDiags := g.rewriteTernaries(expr, g.ternaryTempSpiller)
 	expr, jTemps, jsonDiags := g.rewriteToJSON(expr)

--- a/pkg/codegen/go/gen_program_inline_invoke.go
+++ b/pkg/codegen/go/gen_program_inline_invoke.go
@@ -1,0 +1,78 @@
+package gen
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model"
+	"github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/syntax"
+	"github.com/pulumi/pulumi/pkg/v3/codegen/pcl"
+)
+
+type inlineInvokeTemp struct {
+	Name  string
+	Value *model.FunctionCallExpression
+}
+
+func (temp *inlineInvokeTemp) Type() model.Type {
+	return temp.Value.Type()
+}
+
+func (temp *inlineInvokeTemp) Traverse(traverser hcl.Traverser) (model.Traversable, hcl.Diagnostics) {
+	return temp.Type().Traverse(traverser)
+}
+
+func (temp *inlineInvokeTemp) SyntaxNode() hclsyntax.Node {
+	return syntax.None
+}
+
+type inlineInvokeSpiller struct {
+	temps []*inlineInvokeTemp
+	count int
+}
+
+func (spiller *inlineInvokeSpiller) spillExpression(
+	expr model.Expression,
+	g *generator,
+) (model.Expression, hcl.Diagnostics) {
+	switch expr := expr.(type) {
+	case *model.FunctionCallExpression:
+		isOutputInvoke, _, _ := pcl.RecognizeOutputVersionedInvoke(expr)
+		if expr.Name == "invoke" && !isOutputInvoke {
+			// non-output-versioned invokes are the only ones that need to be converted
+			// because their return type Tuple(InvokeResult, error)
+			_, _, fn, _ := g.functionName(expr.Args[0])
+			tempName := fmt.Sprintf("invoke%s%d", fn, spiller.count)
+			if spiller.count == 0 {
+				tempName = fmt.Sprintf("invoke%s", fn)
+			}
+			temp := &inlineInvokeTemp{
+				Name:  tempName,
+				Value: expr,
+			}
+			spiller.temps = append(spiller.temps, temp)
+
+			spiller.count++
+			reference := model.VariableReference(&model.Variable{
+				Name: temp.Name,
+			})
+
+			return reference, nil
+		}
+
+		return expr, nil
+	default:
+		return expr, nil
+	}
+}
+
+func (g *generator) rewriteInlineInvokes(x model.Expression) (model.Expression, []*inlineInvokeTemp) {
+	spiller := g.inlineInvokeSpiller
+	spiller.temps = nil
+	spill := func(expr model.Expression) (model.Expression, hcl.Diagnostics) {
+		return spiller.spillExpression(expr, g)
+	}
+	expr, _ := model.VisitExpression(x, spill, spill)
+	return expr, spiller.temps
+}

--- a/pkg/codegen/go/gen_program_test.go
+++ b/pkg/codegen/go/gen_program_test.go
@@ -488,6 +488,7 @@ func newTestGenerator(t *testing.T, testFile string) *generator {
 		readDirTempSpiller:  &readDirSpiller{},
 		splatSpiller:        &splatSpiller{},
 		optionalSpiller:     &optionalSpiller{},
+		inlineInvokeSpiller: &inlineInvokeSpiller{},
 		scopeTraversalRoots: codegen.NewStringSet(),
 		arrayHelpers:        make(map[string]*promptToInputArrayHelper),
 		importer:            newFileImporter(),

--- a/pkg/codegen/testing/test/program_driver.go
+++ b/pkg/codegen/testing/test/program_driver.go
@@ -383,6 +383,11 @@ var PulumiPulumiProgramTests = []ProgramTest{
 		Description: "Regression test for rewriting qoutes in python",
 		Skip:        allProgLanguages.Except("python"),
 	},
+	{
+		Directory:   "inline-invokes",
+		Description: "Tests whether using inline invoke expressions works",
+		SkipCompile: codegen.NewStringSet("go"),
+	},
 }
 
 var PulumiPulumiYAMLProgramTests = []ProgramTest{

--- a/pkg/codegen/testing/test/testdata/inline-invokes-pp/dotnet/inline-invokes.cs
+++ b/pkg/codegen/testing/test/testdata/inline-invokes-pp/dotnet/inline-invokes.cs
@@ -1,0 +1,17 @@
+using System.Collections.Generic;
+using System.Linq;
+using Pulumi;
+using Aws = Pulumi.Aws;
+
+return await Deployment.RunAsync(() => 
+{
+    var webSecurityGroup = new Aws.Ec2.SecurityGroup("webSecurityGroup", new()
+    {
+        VpcId = Aws.Ec2.GetVpc.Invoke(new()
+        {
+            Default = true,
+        }).Apply(invoke => invoke.Id),
+    });
+
+});
+

--- a/pkg/codegen/testing/test/testdata/inline-invokes-pp/go/inline-invokes.go
+++ b/pkg/codegen/testing/test/testdata/inline-invokes-pp/go/inline-invokes.go
@@ -1,0 +1,24 @@
+package main
+
+import (
+	"github.com/pulumi/pulumi-aws/sdk/v5/go/aws/ec2"
+	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
+)
+
+func main() {
+	pulumi.Run(func(ctx *pulumi.Context) error {
+		invokeLookupVpc, err := ec2.LookupVpc(ctx, &ec2.LookupVpcArgs{
+			Default: pulumi.BoolRef(true),
+		}, nil)
+		if err != nil {
+			return err
+		}
+		_, err = ec2.NewSecurityGroup(ctx, "webSecurityGroup", &ec2.SecurityGroupArgs{
+			VpcId: invokeLookupVpc.Id,
+		})
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+}

--- a/pkg/codegen/testing/test/testdata/inline-invokes-pp/inline-invokes.pp
+++ b/pkg/codegen/testing/test/testdata/inline-invokes-pp/inline-invokes.pp
@@ -1,0 +1,3 @@
+resource webSecurityGroup "aws:ec2:SecurityGroup" {
+  vpcId = invoke("aws:ec2:getVpc", { default = true }).id
+}

--- a/pkg/codegen/testing/test/testdata/inline-invokes-pp/nodejs/inline-invokes.ts
+++ b/pkg/codegen/testing/test/testdata/inline-invokes-pp/nodejs/inline-invokes.ts
@@ -1,0 +1,6 @@
+import * as pulumi from "@pulumi/pulumi";
+import * as aws from "@pulumi/aws";
+
+const webSecurityGroup = new aws.ec2.SecurityGroup("webSecurityGroup", {vpcId: aws.ec2.getVpc({
+    "default": true,
+}).then(invoke => invoke.id)});

--- a/pkg/codegen/testing/test/testdata/inline-invokes-pp/python/inline-invokes.py
+++ b/pkg/codegen/testing/test/testdata/inline-invokes-pp/python/inline-invokes.py
@@ -1,0 +1,4 @@
+import pulumi
+import pulumi_aws as aws
+
+web_security_group = aws.ec2.SecurityGroup("webSecurityGroup", vpc_id=aws.ec2.get_vpc(default=True).id)

